lib/rouge/lexers/ada.rb

```diff
@@ -0,0 +1,146 @@
+
+          %w(
+            boolean character constraint_error duration float integer
+            natural positive long_float long_integer long_long_float
+            long_long_integer program_error short_float short_integer
+            short_short_integer storage_error string tasking_error
+            wide_character wide_string wide_wide_character
+            wide_wide_string
+          ).each {|w| h[w] = Name::Builtin}
+        end
+      end
+
+      state :whitespace do
+        rule %r{\s+}m, Text
+        rule %r{--.*$}, Comment::Single
+      end
+
+      state :dquote_string do
+        rule %r{[^"\n]+}, Literal::String::Double
+        rule %r{""}, Literal::String::Escape
+        rule %r{"}, Literal::String::Double, :pop!
+        rule %r{\n}, Error, :pop!
+      end
+
+      state :attr do
+        mixin :whitespace
+        rule ID, Name::Attribute, :pop!
+        rule %r{}, Text, :pop!
+      end
+
+      # Handle a dotted name immediately following a declaration keyword.
+      state :decl_name do
+        mixin :whitespace
+        rule %r{body\b}i, Keyword::Declaration  # package body Foo.Bar is...
+        rule %r{(#{ID})(\.)} do
+          groups Name::Namespace, Punctuation
+        end
+        # function "<=" (Left, Right: Type) is ...
+        rule %r{#{ID}|"(and|or|xor|/?=|<=?|>=?|\+|–|&\|/|mod|rem|\*?\*|abs|not)"},
+             Name::Function, :pop!
+        rule %r{}, Text, :pop!
+      end
+
+      # Handle a sequence of library unit names: with Ada.Foo, Ada.Bar;
+      state :libunit_name do
+        mixin :whitespace
+        rule ID do
+          token Name::Namespace
+          goto :libunit_sep
+        end
+        rule %r{}, Text, :pop!
+      end
+
+      state :libunit_sep do
+        mixin :whitespace
+        rule %r{[.,]} do
+          token Punctuation
+          goto :libunit_name
+        end
+        rule %r{}, Text, :pop!
+      end
+
+      state :root do
+        mixin :whitespace
+
+        # String literals.
+        rule %r{'.'}, Literal::String::Char
+        rule %r{"[^"\n]*}, Literal::String::Double, :dquote_string
+
+        # Real literals.
+        rule %r{#{NUM}\.#{NUM}#{EXP}}, Literal::Number::Float
+        rule %r{#{NUM}##{XNUM}\.#{XNUM}##{EXP}}, Literal::Number::Float
+
+        # Integer literals.
+        rule %r{2#[01](_?[01])*##{EXP}}, Literal::Number::Bin
+        rule %r{8#[0-7](_?[0-7])*##{EXP}}, Literal::Number::Oct
+        rule %r{16##{XNUM}*##{EXP}}, Literal::Number::Hex
+        rule %r{#{NUM}##{XNUM}##{EXP}}, Literal::Number::Integer
+        rule %r{#{NUM}#\w+#}, Error
+        rule %r{#{NUM}#{EXP}}, Literal::Number::Integer
+
+        # Special constructs.
+        rule %r{'}, Punctuation, :attr
+        rule %r{<<#{ID}>>}, Name::Label
+
+        # Context clauses are tricky because the 'with' keyword is used
+        # for many purposes. Detect at beginning of the line only.
+        rule %r{^(limited\s+)?(private\s+)?with\b}i, Keyword::Namespace, :libunit_name
```

Comment:
It's not going to be visible in most instances of the output because it's whitespace but this tokenises whitespace with the `Keyword::Namespace` token. It's not as clean as a one-liner but I'd suggest the following:

```ruby
rule %r{^(?:(limited)(\s+))?(?:(private)(\s+))?(with)\b}i do |m|
  groups Keyword::Namespace, Text::Whitespace, Keyword::Namespace, Text::Whitespace, Keyword::Namespace
  push :libunit_name
end
```

Reply:
You caught me! This was on my mind when I wrote that rule, but I got lazy. I'll fix it.

If you don't mind, I'll use plain `Text` instead of `Text::Whitespace` for the whitespace because that's what my `:whitespace` rule does. I copied this behavior from other lexers. I'm not quite sure what `Text::Whitespace` is for—the pygments documentation says *"for specially highlighted whitespace"* and *"The Whitespace token type is new in Pygments 0.8. It is used only by the VisibleWhitespaceFilter currently."*

Reply:
@stoklund No worries. Yeah, I don't think it really has a special meaning in Rouge. As you said, plenty of lexers just use `Text` so that's fine.